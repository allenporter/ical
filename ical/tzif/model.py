"""Data model for the tzif library."""

from collections import namedtuple
from dataclasses import dataclass
from typing import Optional

from .tz_rule import Rule


@dataclass
class Transition:
    """An individual item in the Datablock."""

    transition_time: int
    """A transition time at which the rules for computing local time may change."""

    utoff: int
    """Number of seconds added to UTC to determine local time."""

    dst: bool
    """Determines if local time is Daylight Savings Time (else Standard time)."""

    isstdcnt: bool
    """Determines if the transition time is standard time (else, wall clock time)."""

    isutccnt: bool
    """Determines if the transition time is UTC time, else is a local time."""

    designation: str
    """A designation string."""


LeapSecond = namedtuple("LeapSecond", ["occurrence", "correction"])
"""A correction that needs to be applied to UTC in order to determine TAI.

The occurrence is the time at which the leap-second correction occurs.
The correction is the value of LEAPCORR on or after the occurrence (1 or -1).
"""


@dataclass
class TimezoneInfo:
    """The results of parsing the TZif file."""

    transitions: list[Transition]
    """Local time changes."""

    leap_seconds: list[LeapSecond]

    rule: Optional[Rule] = None
    """A rule for computing local time changes after the last transition."""
