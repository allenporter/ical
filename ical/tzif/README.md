# tzif

This python library provides time transitions for timezones. This library does
not directly provide time zone data, and instead uses the same sources as
`zoneinfo` which includes the system timezone database or the `tzdata` package.

## Background

iCalendar files are fully specified, and there is no ambiguity with respect
to how times in timezones are interpreted. That is, an iCalendar contains a
VTIMEZONE that has all local time transitions fully specified. That means
that an iCalendar implementation must have a timezone database.

[PEP 615](https://peps.python.org/pep-0615/) adds support for the IANA Time Zone
Database in the base python and describes the datasources used to implement
timezones and the `zoneinfo` package. However, those APIs do not expose all
the underlying timezone transitions.

## Details

This package works similarly to the `zoneinfo` package, exposing the underlying
datasources for use by libraries that need timezone transitions. Timezone data
is stored in the Time Zone Information Format (TZif) described in [rfc8536](https://datatracker.ietf.org/doc/html/rfc8536).

The library will read both from the system TZ paths or from the tzdata package.

TZif v3 files support a posix TZ string that describe the rules for the timezone
going forward without explicit transitions. These two resources were helpful to
better describe the string since it is not references in the RFC for TZif:

   - https://developer.ibm.com/articles/au-aix-posix/
   - https://www.di-mgt.com.au/wclock/help/wclo_tzexplain.html
