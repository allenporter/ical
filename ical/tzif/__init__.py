"""Library for parsing python tzdata."""
