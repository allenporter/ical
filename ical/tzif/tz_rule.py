"""Library for parsing TZ rules.

TZ supports these two formats

No DST: std offset
  - std: Name of the timezone
  - offset: Time added to local time to get UTC
  Example: EST+5

DST: std offset dst [offset],start[/time],end[/time]
  - dst: Name of the Daylight savings time timezone
  - offset: Defaults to 1 hour ahead of STD offset if not specified
  - start & end: Time period when DST is in effect. The start/end have
    the following formats:
      Jn: A julian day between 1 and 365 (Feb 29th never counted)
      n: A julian day between 0 and 364 (Feb 29th is counted in leap years)
      Mm.w.d:
          m: Month between 1 and 12
          d: Between 0 (Sunday) and 6 (Saturday)
          w: Between 1 and 5. Week 1 is first week d occurs
      The time field is in hh:mm:ss. The hour can be 167 to -167.
"""

from __future__ import annotations

import datetime
import logging
from typing import Any, Optional, Union, cast

from dateutil import rrule
try:
    from pydantic.v1 import BaseModel, root_validator, validator
except ImportError:
    from pydantic import BaseModel, root_validator, validator
from pyparsing import (
    Char,
    Combine,
    Group,
    Opt,
    ParseException,
    ParserElement,
    Word,
    alphas,
    nums,
)

_LOGGER = logging.getLogger(__name__)

_ZERO = datetime.timedelta(seconds=0)


def _parse_time(values: Any) -> int | str | datetime.timedelta:
    """Convert an offset from [+/-]hh[:mm[:ss]] to a valid timedelta pydantic format.

    The parse tree dict expects fields of hour, minutes, seconds (see tz_time rule in parser).
    """
    if not values:
        return 0
    if isinstance(values, (int, str, datetime.timedelta)):
        return values
    if not isinstance(values, dict):
        raise ValueError("time was not parse tree dict, timedelta, string, or int")
    hour = values["hour"]
    sign = 1
    if hour.startswith("+"):
        hour = hour[1:]
    elif hour.startswith("-"):
        sign = -1
        hour = hour[1:]
    minutes = values.get("minutes", "0")
    seconds = values.get("seconds", "0")
    return sign * (int(hour) * 60 * 60 + int(minutes) * 60 + int(seconds))


class RuleDay(BaseModel):
    """A date referenced in a timezone rule for a julian day."""

    day_of_year: int
    """A day of the year between 1 and 365, leap days never supported."""

    time: datetime.timedelta
    """Offset of time in current local time when the rule goes into effect, default of 02:00:00."""

    _parse_time = validator("time", pre=True, allow_reuse=True)(_parse_time)


class RuleDate(BaseModel):
    """A date referenced in a timezone rule."""

    month: int
    """A month between 1 and 12."""

    day_of_week: int
    """A day of the week between 0 (Sunday) and 6 (Saturday)."""

    week_of_month: int
    """A week number of the month (1 to 5) based on the first occurrence of day_of_week."""

    time: datetime.timedelta
    """Offset of time in current local time when the rule goes into effect, default of 02:00:00."""

    _parse_time = validator("time", pre=True, allow_reuse=True)(_parse_time)

    def as_rrule(self, dtstart: datetime.datetime | None = None) -> rrule.rrule:
        """Return a recurrence rule for this timezone occurrence (no start date)."""
        dst_start_weekday = self._rrule_byday(self._rrule_week_of_month)
        if dtstart:
            dtstart = dtstart.replace(hour=0, minute=0, second=0) + self.time
        return rrule.rrule(
            freq=rrule.YEARLY,
            bymonth=self.month,
            byweekday=dst_start_weekday,
            dtstart=dtstart,
        )

    @property
    def rrule_str(self) -> str:
        """Return a recurrence rule string for this timezone occurrence."""
        return ";".join(
            [
                "FREQ=YEARLY",
                f"BYMONTH={self.month}",
                f"BYDAY={self._rrule_week_of_month}{self._rrule_byday}",
            ]
        )

    def rrule_dtstart(self, start: datetime.datetime) -> datetime.datetime:
        """Return an rrule dtstart starting at the specified date with the time applied."""
        dt_start = start.replace(hour=0, minute=0, second=0) + self.time
        return cast(datetime.datetime, next(iter(self.as_rrule(dt_start))))

    @property
    def _rrule_byday(self) -> rrule.weekday:
        """Return the dateutil weekday for this rule based on day_of_week."""
        return rrule.weekdays[(self.day_of_week - 1) % 7]

    @property
    def _rrule_week_of_month(self) -> int:
        """Return the byday modifier for the week of the month."""
        if self.week_of_month == 5:
            return -1
        return self.week_of_month


class RuleOccurrence(BaseModel):
    """A TimeZone rule occurrence."""

    name: str
    """The name of the timezone occurrence e.g. EST."""

    offset: datetime.timedelta
    """UTC offset for this timezone occurrence (not time added to local time)."""

    _parse_offset = validator("offset", pre=True, allow_reuse=True)(_parse_time)

    @validator("offset", allow_reuse=True)
    def negate_offset(cls, value: datetime.timedelta) -> datetime.timedelta:
        """Convert the offset from time added to local time to get UTC to a UTC offset."""
        result = _ZERO - value
        return result


def _default_time_value(values: dict[str, Any]) -> dict[str, Any]:
    """Set a default time value when none is specified."""
    if "time" not in values:
        values["time"] = {"hour": "2"}
    return values


class Rule(BaseModel):
    """A rule for evaluating future timezone transitions."""

    std: RuleOccurrence
    """An occurrence of a timezone transition for standard time."""

    dst: Optional[RuleOccurrence] = None
    """An occurrence of a timezone transition for standard time."""

    dst_start: Union[RuleDate, RuleDay, None] = None
    """Describes when dst goes into effect."""

    dst_end: Union[RuleDate, RuleDay, None] = None
    """Describes when dst ends (std starts)."""

    _default_start_time = validator("dst_start", pre=True, allow_reuse=True)(
        _default_time_value
    )
    _default_end_time = validator("dst_end", pre=True, allow_reuse=True)(
        _default_time_value
    )

    @root_validator(allow_reuse=True)
    def default_dst_offset(cls, values: dict[str, Any]) -> dict[str, Any]:
        """Infer the default DST offset based on STD offset if not specified."""
        if values.get("dst") and not values["dst"].offset:
            # If the dst offset is omitted, it defaults to one hour ahead of standard time.
            values["dst"].offset = values["std"].offset + datetime.timedelta(hours=1)
        return values


def parse_tz_rule(tz_str: str) -> Rule:
    """Parse the TZ string into a Rule object."""

    hour = Combine(Opt(Word("+-")) + Word(nums))
    tz_time = hour.set_results_name("hour") + Opt(
        ":"
        + Word(nums).set_results_name("minutes")
        + Opt(":" + Word(nums).set_results_name("seconds"))
    )

    name: ParserElement
    # Hack for inability to deal with both start word options
    if tz_str.startswith("<"):
        name = Combine(Char("<") + Opt(Word("+-")) + Word(nums) + Char(">"))
    else:
        name = Word(alphas)

    onset = name.set_results_name("name") + Group(Opt(tz_time)).set_results_name(
        "offset"
    )
    month_date = (
        "M"
        + Word(nums).set_results_name("month")
        + "."
        + Word(nums).set_results_name("week_of_month")
        + "."
        + Word(nums).set_results_name("day_of_week")
    )
    julian_date = "J" + Word(nums).set_results_name("day_of_year")
    # Hack for inabiliy to have a single rule with both date types
    if ",J" in tz_str:
        tz_days = julian_date
    else:
        tz_days = month_date
    tz_date = tz_days + Opt("/" + Group(tz_time).set_results_name("time"))

    tz_rule = (
        Group(onset).set_results_name("std")
        + Opt(Group(onset).set_results_name("dst"))
        + Opt(
            ","
            + Group(tz_date).set_results_name("dst_start")
            + ","
            + Group(tz_date).set_results_name("dst_end")
        )
    )
    if _LOGGER.isEnabledFor(logging.DEBUG):
        tz_rule.set_debug(flag=True)
    try:
        result = tz_rule.parse_string(tz_str, parse_all=True)
    except ParseException as err:
        raise ValueError(f"Unable to parse TZ string: {tz_str}") from err

    return Rule.parse_obj(result.as_dict())
