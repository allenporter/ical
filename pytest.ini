[pytest]
log_level = DEBUG
# Can be run manually with --benchmark-only
addopts = --benchmark-skip
