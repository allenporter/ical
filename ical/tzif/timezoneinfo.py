"""Library for returning details about a timezone.

This package follows the same approach as zoneinfo for loading timezone
data. It first checks the system TZPATH, then falls back to the tzdata
python package.
"""

from __future__ import annotations

import datetime
import logging
import os
import zoneinfo
from dataclasses import dataclass
from functools import cache
from importlib import resources

from .model import TimezoneInfo
from .tz_rule import Rule, RuleDate
from .tzif import read_tzif

_LOGGER = logging.getLogger(__name__)


class TimezoneInfoError(Exception):
    """Raised on error working with timezone information."""


@cache
def _read_system_timezones() -> set[str]:
    """Read and cache the set of system and tzdata timezones."""
    return zoneinfo.available_timezones()


def _find_tzfile(key: str) -> str | None:
    """Retrieve the path to a TZif file from a key."""
    for search_path in zoneinfo.TZPATH:
        filepath = os.path.join(search_path, key)
        if os.path.isfile(filepath):
            return filepath

    return None


@cache
def _read_tzdata_timezones() -> set[str]:
    """Returns the set of valid timezones from tzdata only."""
    try:
        with resources.files("tzdata").joinpath("zones").open(
            "r", encoding="utf-8"
        ) as zones_file:
            return {line.strip() for line in zones_file.readlines()}
    except ModuleNotFoundError:
        return set()


def _iana_key_to_resource(key: str) -> tuple[str, str]:
    """Returns the package and resource file for the specified timezone."""
    if "/" not in key:
        return "tzdata.zoneinfo", key
    package_loc, resource = key.rsplit("/", 1)
    package = "tzdata.zoneinfo." + package_loc.replace("/", ".")
    return package, resource


def read(key: str) -> TimezoneInfo:
    """Read the TZif file from the tzdata package and return timezone records."""
    if key not in _read_system_timezones() and key not in _read_tzdata_timezones():
        raise TimezoneInfoError(f"Unable to find timezone in system timezones: {key}")

    # Prefer tzdata package
    (package, resource) = _iana_key_to_resource(key)
    try:
        with resources.files(package).joinpath(resource).open("rb") as tzdata_file:
            return read_tzif(tzdata_file.read())
    except ModuleNotFoundError:
        # Unexpected given we previously read the list of timezones
        pass
    except FileNotFoundError as err:
        raise TimezoneInfoError(f"Unable to load tzdata module: {key}") from err

    # Fallback to zoneinfo file on local disk
    tzfile = _find_tzfile(key)
    if tzfile is not None:
        with open(tzfile, "rb") as tzfile_file:
            return read_tzif(tzfile_file.read())

    raise TimezoneInfoError(f"Unable to find timezone data for {key}")


_ZERO = datetime.timedelta(0)
_HOUR = datetime.timedelta(hours=1)


@dataclass
class TzInfoResult:
    """Contains timezone information for a specific datetime."""

    utcoffset: datetime.timedelta
    tzname: str | None
    dst: datetime.timedelta | None


class TzInfo(datetime.tzinfo):
    """An implementation of tzinfo based on a TimezoneInfo for current TZ rules.

    This class is not as complete of an implementation of pythons zoneinfo rules as
    it currently ignores historical timezone information. Instead, it uses only the
    posix TZ rules that apply going forward only, but applies them for all time.

    This class uses the default implementation of fromutc.
    """

    def __init__(self, rule: Rule) -> None:
        """Initialize TzInfo."""
        self._rule: Rule = rule

    @classmethod
    def from_timezoneinfo(cls, timezoneinfo: TimezoneInfo) -> TzInfo:
        """Create a new instance of a TzInfo."""
        if not timezoneinfo.rule:
            raise ValueError("Unable to make TzInfo from TimezoneInfo, missing rule")
        return cls(timezoneinfo.rule)

    def utcoffset(self, dt: datetime.datetime | None) -> datetime.timedelta:
        """Return offset of local time from UTC, as a timedelta object."""
        if not dt:
            return _ZERO
        result = self._rule.std.offset
        if dst_offset := self.dst(dt):
            result += dst_offset
        return result

    def tzname(self, dt: datetime.datetime | None) -> str | None:
        """Return the time zone name for the datetime as a sorting."""
        if dt is None:
            return None
        if self.dst(dt) and self._rule.dst:
            return self._rule.dst.name
        return self._rule.std.name

    def dst(self, dt: datetime.datetime | None) -> datetime.timedelta | None:
        """Return the daylight saving time (DST) adjustment, if applicable."""
        if (
            dt is None
            or not self._rule.dst
            or not isinstance(self._rule.dst_start, RuleDate)
            or not isinstance(self._rule.dst_end, RuleDate)
            or not self._rule.dst.offset
        ):
            return None

        dt_year = datetime.datetime(dt.year, 1, 1)
        dst_start = next(iter(self._rule.dst_start.as_rrule(dt_year)))
        dst_end = next(iter(self._rule.dst_end.as_rrule(dt_year)))
        if dst_start <= dt.replace(tzinfo=None) < dst_end:
            dst_offset = self._rule.dst.offset - self._rule.std.offset
            return dst_offset

        return _ZERO


def read_tzinfo(key: str) -> TzInfo:
    """Create a zoneinfo implementation from raw tzif data."""
    timezoneinfo = read(key)
    try:
        return TzInfo.from_timezoneinfo(timezoneinfo)
    except ValueError as err:
        raise TimezoneInfoError(f"Unable create TzInfo: {key}") from err
