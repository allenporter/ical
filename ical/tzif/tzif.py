"""Library for reading python tzdata files.

An rfc5545 calendar references timezones in an unambiguous way, defined
by a set of specific transitions. As a result, a calendar must have
complete information about the timezones it references.

Python's zoneinfo package does not expose a full representation of system
supported timezones. It will use the system defined timezone, but then also
may fallback to the tzdata package as a fallback.

This package uses the tzdata package as a definition for timezones, under
the assumption that it should be similar to existing python supported
timezones.

Note: This implementation is more verbose than the zoneinfo implementation
and contains more documentation and references to the file format to serve
as a resource for understanding the format. See rfc8536 for TZif file format.
"""

import enum
import io
import logging
import struct
from collections import namedtuple
from collections.abc import Callable
from dataclasses import dataclass
from functools import cache
from typing import Sequence

from .model import LeapSecond, TimezoneInfo, Transition
from .tz_rule import parse_tz_rule

_LOGGER = logging.getLogger(__name__)

# Records specifying the local time type
_LOCAL_TIME_TYPE_STRUCT_FORMAT = "".join(
    [
        ">",  # Use standard size of packed value bytes
        "l",  # utoff (4 bytes): Number of seconds to add to UTC to determine local time
        "?",  # dst (1 byte): Indicates the time is DST (1) or standard (0)
        "B",  # idx (1 byte): Offset index into the time zone designiation octets (0-charcnt-1)
    ]
)
_LOCAL_TIME_RECORD_SIZE = 6


class _TZifVersion(enum.Enum):
    """Defines information related to _TZifVersions."""

    V1 = (b"\x00", 4, "l")  # 32-bit in v1
    V2 = (b"2", 8, "q")  # 64-bit in v2+
    V3 = (b"3", 8, "q")

    def __init__(self, version: bytes, time_size: int, time_format: str):
        self._version = version
        self._time_size = time_size
        self._time_format = time_format

    @property
    def version(self) -> bytes:
        """Return the version byte string."""
        return self._version

    @property
    def time_size(self) -> int:
        """Return the TIME_SIZE used in the data block parsing."""
        return self._time_size

    @property
    def time_format(self) -> str:
        """Return the struct unpack format string for TIME_SIZE objects."""
        return self._time_format


@dataclass
class _Header:
    """TZif _Header information."""

    SIZE = 44  # Total size of the header to read
    STRUCT_FORMAT = "".join(
        [
            ">",  # Use standard size of packed value bytes
            "4s",  # magic (4 bytes)
            "c",  # version (1 byte)
            "15x",  # unused
            "6l",  # isutccnt, isstdcnt, leapcnt, timecnt, typecnt, charcnt
        ]
    )
    MAGIC = "TZif".encode()

    version: bytes
    """The version of the files format."""

    isutccnt: int
    """The number of UTC/local indicators in the data block."""

    isstdcnt: int
    """The number of standard/wall indicators in the data block."""

    leapcnt: int
    """The number of leap second records in the data block."""

    timecnt: int
    """The number of time transitions in the data block."""

    typecnt: int
    """The number of local time type records in the data block."""

    charcnt: int
    """The number of characters for time zone designations in the data block."""

    @classmethod
    def from_bytes(cls, header_bytes: bytes) -> "_Header":
        """Parse the header bytes into a file."""
        (
            magic,
            version,
            isutccnt,
            isstdcnt,
            leapcnt,
            timecnt,
            typecnt,
            charcnt,
        ) = struct.unpack(_Header.STRUCT_FORMAT, header_bytes)
        if magic != _Header.MAGIC:
            raise ValueError("zoneinfo file did not contain magic header")
        if isutccnt not in (0, typecnt):
            raise ValueError(
                f"UTC/local indicators in datablock mismatched ({isutccnt}, {typecnt})"
            )
        if isstdcnt not in (0, typecnt):
            raise ValueError(
                f"standard/wall indicators in datablock mismatched ({isutccnt}, {typecnt})"
            )
        return _Header(version, isutccnt, isstdcnt, leapcnt, timecnt, typecnt, charcnt)


_TransitionBlock = namedtuple(
    "_TransitionBlock", ["transition_time", "time_type", "isstdcnt", "isutccnt"]
)

# A series of records specifying the local time type:
#  - utoff (4 bytes): Number of seconds to add to UTC to determine local time
#  - dst (1 byte): Indicates the time is DST (1) or standard (0)
#  - idx (1 byte):  Offset index into the time zone designiation octets (0-charcnt-1)
# is the utoff (4 bytes), dst (1 byte), idx (1 byte).
_LocalTimeType = namedtuple("_LocalTimeType", ["utoff", "dst", "idx"])


def _new_transition(
    transition: _TransitionBlock,
    local_time_types: list[_LocalTimeType],
    get_tz_designations: Callable[[int], str],
) -> Transition:
    """ddd."""
    if transition.time_type >= len(local_time_types):
        raise ValueError(
            f"transition_type out of bounds {transition.time_type} >= {len(local_time_types)}"
        )
    if transition.isutccnt and not transition.isstdcnt:
        raise ValueError("isutccnt was True but isstdcnt was False")
    (utoff, dst, idx) = local_time_types[transition.time_type]
    return Transition(
        transition.transition_time,
        utoff,
        dst,
        transition.isstdcnt,
        transition.isutccnt,
        get_tz_designations(idx),
    )


def _read_datablock(
    header: _Header, version: _TZifVersion, buf: io.BytesIO
) -> tuple[list[Transition], list[LeapSecond]]:
    """Read records from the buffer."""
    # A series of leap-time values in sorted order
    transition_times = struct.unpack(
        f">{header.timecnt}{version.time_format}",
        buf.read(header.timecnt * version.time_size),
    )

    # A series of integers specifying the type of local time of the corresponding
    # transition time. These are zero-based indices into the array of local
    # time type records. (from 0 to typecnt-1)
    transition_types: Sequence[int] = []
    if header.timecnt > 0:
        transition_types = struct.unpack(
            f">{header.timecnt}B", buf.read(header.timecnt)
        )

    local_time_types: list[_LocalTimeType] = [
        _LocalTimeType._make(
            struct.unpack(
                _LOCAL_TIME_TYPE_STRUCT_FORMAT, buf.read(_LOCAL_TIME_RECORD_SIZE)
            )
        )
        for _ in range(header.typecnt)
    ]

    # An array of NUL-terminated time zone designation strings
    tz_designations = buf.read(header.charcnt)

    @cache
    def get_tz_designations(idx: int) -> str:
        """Find the null terminated string starting at the specified index."""
        end = tz_designations.find(b"\x00", idx)
        return tz_designations[idx:end].decode("UTF-8")

    leap_seconds: list[LeapSecond] = [
        LeapSecond._make(
            struct.unpack(
                f">{version.time_format}l",
                buf.read(version.time_size + 4),  # occur + corr
            )
        )
        for _ in range(header.leapcnt)
    ]

    # Standard/wall indicators determine if the transition times are standard time (1)
    # or wall clock time (0).
    isstdcnt_types: list[bool] = []
    if header.isstdcnt > 0:
        isstdcnt_types.extend(
            struct.unpack(
                f">{header.isstdcnt}?",
                buf.read(header.isstdcnt),
            )
        )
    isstdcnt_types.extend([False] * (header.timecnt - header.isstdcnt))

    # UTC/local indicators determine if the transition times are UTC (1) or local time (0).
    isutccnt_types: list[bool] = []
    if header.isutccnt > 0:
        isutccnt_types.extend(
            struct.unpack(f">{header.isutccnt}?", buf.read(header.isutccnt))
        )
    isutccnt_types.extend([False] * (header.timecnt - header.isutccnt))

    transitions = [
        _new_transition(
            _TransitionBlock(*values), local_time_types, get_tz_designations
        )
        for values in zip(
            transition_times, transition_types, isstdcnt_types, isutccnt_types
        )
    ]

    return (transitions, leap_seconds)


def read_tzif(content: bytes) -> TimezoneInfo:
    """Read the TZif file and parse and return the timezone records."""
    buf = io.BytesIO(content)

    # V1 header and block
    header = _Header.from_bytes(buf.read(_Header.SIZE))
    if header.version == _TZifVersion.V1.version:
        if header.typecnt == 0:
            raise ValueError("Local time records in block is zero")
        if header.charcnt == 0:
            raise ValueError("Total number of octets is zero")
    (transitions, leap_seconds) = _read_datablock(header, _TZifVersion.V1, buf)
    if header.version == _TZifVersion.V1.version:
        return TimezoneInfo(transitions, leap_seconds)

    # V2+ header and block
    header = _Header.from_bytes(buf.read(_Header.SIZE))
    if header.typecnt == 0:
        raise ValueError("Local time records in block is zero")
    if header.charcnt == 0:
        raise ValueError("Total number of octets is zero")

    (transitions, leap_seconds) = _read_datablock(header, _TZifVersion.V2, buf)

    # V2+ footer
    footer = buf.read()
    parts = footer.decode("UTF-8").split("\n")
    if len(parts) != 3:
        raise ValueError("Failed to read TZ footer")
    rule = None
    if parts[1]:
        rule = parse_tz_rule(parts[1])
    return TimezoneInfo(transitions, leap_seconds, rule=rule)
