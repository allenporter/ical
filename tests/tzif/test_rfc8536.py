"""Tests for rfc8536 examples."""

import dataclasses
import io
import re

import pytest
from pytest_golden.plugin import GoldenTestFixture

from ical.tzif.tzif import read_tzif

RFC_LINE = re.compile(r"\|(?:[0-9]|\s)+\| (.*?)\s+\| .+ \| .+ \|")


def rfc_to_binary(rfc_text: str) -> bytes:
    """Convert the RFC example text to a binary blob."""
    buf = io.BytesIO()
    for line in rfc_text.split("\n"):
        match = RFC_LINE.match(line)
        if not match:
            continue
        if not (payload := match.group(1)):
            continue
        buf.write(bytearray.fromhex(payload.replace(" ", "")))
    return buf.getvalue()


@pytest.mark.golden_test("testdata/rfc*.yaml")
def test_parse(golden: GoldenTestFixture) -> None:
    """Test that reads RFC examples from golden files."""
    if golden.get("disabled"):
        return

    content = rfc_to_binary(golden["input"])
    result = read_tzif(content)
    assert dataclasses.asdict(result) == golden.out["output"]
