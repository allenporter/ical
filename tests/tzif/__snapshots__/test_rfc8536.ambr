# serializer version: 1
# name: test_parse[rfc8536-v1]
  dict({
    'leap_seconds': list([
      list([
        78796800,
        1,
      ]),
      list([
        94694401,
        2,
      ]),
      list([
        126230402,
        3,
      ]),
      list([
        157766403,
        4,
      ]),
      list([
        189302404,
        5,
      ]),
      list([
        220924805,
        6,
      ]),
      list([
        252460806,
        7,
      ]),
      list([
        283996807,
        8,
      ]),
      list([
        315532808,
        9,
      ]),
      list([
        362793609,
        10,
      ]),
      list([
        394329610,
        11,
      ]),
      list([
        425865611,
        12,
      ]),
      list([
        489024012,
        13,
      ]),
      list([
        567993613,
        14,
      ]),
      list([
        631152014,
        15,
      ]),
      list([
        662688015,
        16,
      ]),
      list([
        709948816,
        17,
      ]),
      list([
        741484817,
        18,
      ]),
      list([
        773020818,
        19,
      ]),
      list([
        820454419,
        20,
      ]),
      list([
        867715220,
        21,
      ]),
      list([
        915148821,
        22,
      ]),
      list([
        1136073622,
        23,
      ]),
      list([
        1230768023,
        24,
      ]),
      list([
        1341100824,
        25,
      ]),
      list([
        1435708825,
        26,
      ]),
      list([
        1483228826,
        27,
      ]),
    ]),
  })
# ---
# name: test_parse[rfc8536-v2]
  dict({
    'rule': dict({
      'dst': None,
      'dst_end': None,
      'dst_start': None,
      'std': dict({
        'name': 'HST',
        'offset': '-PT10H',
      }),
    }),
    'transitions': list([
      dict({
        'designation': 'HST',
        'dst': False,
        'isstdcnt': False,
        'isutccnt': False,
        'transition_time': -2334101314,
        'utoff': -37800,
      }),
      dict({
        'designation': 'HDT',
        'dst': True,
        'isstdcnt': False,
        'isutccnt': False,
        'transition_time': -1157283000,
        'utoff': -34200,
      }),
      dict({
        'designation': 'HST',
        'dst': False,
        'isstdcnt': False,
        'isutccnt': False,
        'transition_time': -1155436200,
        'utoff': -37800,
      }),
      dict({
        'designation': 'HWT',
        'dst': True,
        'isstdcnt': False,
        'isutccnt': False,
        'transition_time': -880198200,
        'utoff': -34200,
      }),
      dict({
        'designation': 'HPT',
        'dst': True,
        'isstdcnt': True,
        'isutccnt': True,
        'transition_time': -769395600,
        'utoff': -34200,
      }),
      dict({
        'designation': 'HST',
        'dst': False,
        'isstdcnt': False,
        'isutccnt': False,
        'transition_time': -765376200,
        'utoff': -37800,
      }),
      dict({
        'designation': 'HST',
        'dst': False,
        'isstdcnt': False,
        'isutccnt': False,
        'transition_time': -712150200,
        'utoff': -36000,
      }),
    ]),
  })
# ---
